ovs-p4rt/ovs_p4rt.cc

```diff
@@ -278,10 +278,115 @@ void PrepareFdbRxVlanTableEntry(p4::v1::TableEntry* table_entry,
 }
 #endif
 
-void PrepareFdbTableEntryforV4Tunnel(p4::v1::TableEntry* table_entry,
-                                     const struct mac_learning_info& learn_info,
-                                     const ::p4::config::v1::P4Info& p4info,
-                                     bool insert_entry) {
+void PrepareFdbTableEntryforV4VxlanTunnel(
+    p4::v1::TableEntry* table_entry, const struct mac_learning_info& learn_info,
+    const ::p4::config::v1::P4Info& p4info, bool insert_entry) {
+  table_entry->set_table_id(GetTableId(p4info, L2_FWD_TX_TABLE));
+  auto match = table_entry->add_match();
+  match->set_field_id(
+      GetMatchFieldId(p4info, L2_FWD_TX_TABLE, L2_FWD_TX_TABLE_KEY_DST_MAC));
+
+  std::string mac_addr = CanonicalizeMac(learn_info.mac_addr);
+  match->mutable_exact()->set_value(mac_addr);
+#if defined(ES2K_TARGET)
+  // Based on p4 program for ES2K, we need to provide a match key Bridge ID
+  auto match1 = table_entry->add_match();
+  match1->set_field_id(
+      GetMatchFieldId(p4info, L2_FWD_TX_TABLE, L2_FWD_TX_TABLE_KEY_BRIDGE_ID));
+
+  match1->mutable_exact()->set_value(EncodeByteValue(1, learn_info.bridge_id));
+
+  // Based on p4 program for ES2K, we need to provide a match key SMAC flag
+  auto match2 = table_entry->add_match();
+  match2->set_field_id(GetMatchFieldId(p4info, L2_FWD_TX_TABLE,
+                                       L2_FWD_TX_TABLE_KEY_SMAC_LEARNED));
+
+  match2->mutable_exact()->set_value(EncodeByteValue(1, 1));
+
+#endif
+
+#if defined(DPDK_TARGET)
+  if (insert_entry) {
+    auto table_action = table_entry->mutable_action();
+    auto action = table_action->mutable_action();
+    action->set_action_id(
+        GetActionId(p4info, L2_FWD_TX_TABLE_ACTION_SET_TUNNEL));
+    {
+      auto param = action->add_params();
+      param->set_param_id(GetParamId(p4info, L2_FWD_TX_TABLE_ACTION_SET_TUNNEL,
+                                     ACTION_SET_TUNNEL_PARAM_TUNNEL_ID));
+      param->set_value(EncodeByteValue(1, learn_info.tnl_info.vni));
+    }
+
+    {
+      auto param = action->add_params();
+      param->set_param_id(GetParamId(p4info, L2_FWD_TX_TABLE_ACTION_SET_TUNNEL,
+                                     ACTION_SET_TUNNEL_PARAM_DST_ADDR));
+      std::string ip_address =
+          CanonicalizeIp(learn_info.tnl_info.remote_ip.ip.v4addr.s_addr);
+      param->set_value(ip_address);
+    }
+  }
+#elif defined(ES2K_TARGET)
+  if (insert_entry) {
+    auto table_action = table_entry->mutable_action();
+    auto action = table_action->mutable_action();
+
+    if (learn_info.tnl_info.local_ip.family == AF_INET &&
```

Comment:
There seems to be a section for IPv4 and IPv6 in ES2K_TARGET. Do we not need something similar for DPDK_TARGET too?

Reply:
Could also drop the `v4` in the PrepareFdbTableEntryforV4VxlanTunnel method name maybe?

Reply:
Yeah, the name needs to be changed. Currently, there is no logger as well. There is some cleanup needed in the future.

Reply:
> There seems to be a section for IPv4 and IPv6 in ES2K_TARGET. Do we not need something similar for DPDK_TARGET too?

I believe Ipv6 is not supported for DPDK_TARGET.